File: HackerRank/WeekOfCode/WeekOfCode32/CircularWalk.py
#!/bin/python3

import sys

n, s, t = input().strip().split(' ')
n, s, t = [int(n), int(s), int(t)]
r_0, g, seed, p = input().strip().split(' ')
r_0, g, seed, p = [int(r_0), int(g), int(seed), int(p)]

def fix(x):
    x %= n
    return x if x >= 0 else x + n

def betw(left, right, x):
    c = right - left
    left = fix(left)
    if x < left: x += n
    return x - left <= c

def circularWalk(n, s, t, r0, g, seed, p):
    r = [0 for i in range(0,n)]
    r[0] = r0
    for i in range(1,n):
        r[i] = (r[i-1] * g + seed) % p
    
    left = s
    right = s
    minl = left - r[s]
    maxr = right + r[s]
    
    k=0
    while True:
        if betw(left, right, t): return k
        if left==minl and right==maxr: return -1
        k += 1   
    
        oldl,oldr,left,right = left,right,minl,maxr
        
        for i in range(left, oldl):
            v = r[fix(i)]
            minl, maxr = min(i-v, minl), max(i+v, maxr)
            
        for i in range(oldr+1, right+1):
            v = r[fix(i)]
            minl, maxr = min(i-v, minl), max(i+v, maxr)

result = circularWalk(n, s, t, r_0, g, seed, p)
print(result)

File: HackerRank/WeekOfCode/WeekOfCode32/CircularWalk.cpp
#include <algorithm>
#include <iomanip>
#include <istream>
#include <map>
#include <numeric>
#include <ostream>
#include <set>
#include <sstream>
#include <string>
#include <utility>
#include <vector>
#include <iostream>
using namespace std;
// Powered by caide (code generator, tester, and library code inliner)


typedef long long ll;


const int MAXN = 1000 * 1000 * 10 + 1;

int r[MAXN];

class Solution {
public:

	ll n, s, t;
	ll r0, g, seed, p;

	void solve(istream& in, ostream& out) {

		in >> n >> s >> t;
		in >> r[0] >> g >> seed >> p;

		for (int i= 1; i<n; i++)
		{
			ll ri = 1LL * r[i - 1] * g + seed; 
			r[i] = ri % p;
		}

		ll left = s;
		ll right = s;
		ll maxLeft = left - r[s];
		ll maxRight = right + r[s];

		ll answer = -1;
		for (int k = 0; k <= n; k++)
		{
			if (between(left, right, t))
			{
				answer = k;
				break;
			}

			if (left == maxLeft && right == maxRight)
				break;

			auto oldLeft = left;
			auto oldRight = right;
			left = maxLeft;
			right = maxRight;

			for (ll i = left; i < oldLeft; i++)
			{
				auto v = r[fix(i)];
				maxLeft = min(i - v, maxLeft);
				maxRight = max(i + v, maxRight);
			}

			for (ll i = oldRight + 1; i <= right; i++)
			{
				auto v = r[fix(i)];
				maxLeft = min(i - v, maxLeft);
				maxRight = max(i + v, maxRight);
			}
		}

		out << answer;
	}

	bool between(ll left, ll right, ll x)
	{
		auto count = right - left;
		left = fix(left);
		if (x < left) x += n;
		return x - left <= count;
	}

	ll fix(ll x) { 	return (x % n + n) % n; }
};

void solve(istream& in, ostream& out)
{
	out << setprecision(12);
	Solution solution;
	solution.solve(in, out);
}

#include <fstream>
int main() {
   
    ios_base::sync_with_stdio(0);
    cin.tie(0);

    istream& in = cin;
    ostream& out = cout;

    solve(in, out);
    return 0;
}

